### PR TITLE
Fix crash when H2 client does not set End-of-data bit

### PR DESCRIPTION
**Pull request by shinrich**
Yet another solution to issue #6313.

Unlike PR #6407 and #6401, this PR does not update the H2 logic to send a READ_COMPLETE when the end-of-data bit is set or when the read_vio.ntodo() is 0.  Rather if the user agent never sends the end-of-data bit, the user agent will eventually send a EOS or get timed out and this PR adds logic in the HttpSM to cleanup the server_session without crashing.

While this PR is more spec compliant, our organization may still go with PR #6407 in the short term anyway because we do have user agents that send POST data but do not set the end-of-data bit.  While running with this patch on a production box, the number of H2 POST ERR_CLIENT_ABORTS increased by an order of magnitude over a 15 minute logging interval.  For us tracking down the appropriate app dev teams takes some time, and getting clients to update apps takes even longer.

You'll noticed in this PR, that I added a boolean return value to Http2Stream::update_read.  The idea was that it would return false if ndone > nbytes.  Then the caller could issue a stream error because the client was breaking spec by sending more data than indicated by the content-length.  However, in many cases nbytes is 0 because after the request header is processed, the read_vio nbytes is set to ndone which at that point is 0.  For about an hour of panicked investigation, I thought that @masaori335's post optimization had broken things if the data frame arrived before the Tunnel's do_io_read was all set up.  But in fact all is well.  The logic in HttpSM::do_setup_post_tunnel will copy extra data from the end of the ua_buffer (which is the _request_buffer for Http2Stream) into the tunnel's buffer, so all is well there anyway.

### CODE REVIEW

**Comment by randall** [comment]
[approve ci autest]

File: proxy/http2/Http2Stream.h [hunk 1]
```diff
@@ -124,6 +124,7 @@ class Http2Stream : public ProxyTransaction
   bool has_trailing_header() const;
   void set_request_headers(HTTPHdr &h2_headers);
   MIOBuffer *read_vio_writer() const;
+  bool update_read(int num_written);
```
Comment (masaori335) [inline proxy/http2/Http2Stream.h]: This name looks too generic. Please rename to `update_read_vio_ndone` or something else. Sorry for bothering you but we already have `update_read_request()`.

**Comment by masaori335** [comment]
If I revert all changes of HTTP/2, I could get 408 and no crash. Updating `read_vio.ndone` might introduce or uncover an issue.
Let's narrow down changes of this PR to only HttpSM.cc, any HTTP/2 changes is not necessary here.

File: proxy/http/HttpSM.cc [hunk 2]
```diff
@@ -6860,6 +6867,9 @@ HttpSM::kill_this()
       plugin_tunnel = nullptr;
     }
 
+    if (server_session && server_entry && !server_entry->in_tunnel && server_entry->vc == server_session) {
```
[the remark this answers is not in the document]
Comment (shinrich) [reply proxy/http/HttpSM.cc]: Yes, then could take this one out.  It didn't twig in my tests as far as I could tell (had a Warning in there while testing).

**Comment by masaori335** [comment]
> While this PR is more spec compliant, our organization may still go with PR #6407 in the short term anyway because we do have user agents that send POST data but do not set the end-of-data bit. 
While running with this patch on a production box, the number of H2 POST ERR_CLIENT_ABORTS increased by an order of magnitude over a 15 minute logging interval. For us tracking down the appropriate app dev teams takes some time, and getting clients to update apps takes even longer.

I commented on #6404, I found that ATS ignore "pure END_STREAM" (DATA frame with END_STREAM & length = 0) and VC_EVENT_READ_COMPLETE is not signaled. It's possible that
this bug made some clients aborts. I prefer to polish this PR & land with #6404.

**Comment by shinrich** [comment]
Updated to address @masaori335 comments.  Tested with PR #6404 successfully.  Updating issue #6313 with details.

**Comment by masaori335** [comment]
[approve ci autest]

**Comment by zwoop** [comment]
Cherry-picked to v9.0.x branch.

**Pull request merged**